core/sr-api-macros/tests/runtime_calls.rs

```diff
@@ -70,11 +70,45 @@ fn calling_native_runtime_signature_changed_function() {
 
 #[test]
 fn calling_wasm_runtime_signature_changed_old_function() {
-	let client = test_client::new_with_api_execution_strat(ExecutionStrategy::AlwaysWasm);
+	let client = test_client::new_with_execution_strategy(ExecutionStrategy::AlwaysWasm);
 	let runtime_api = client.runtime_api();
 	let block_id = BlockId::Number(client.info().unwrap().chain.best_number);
 
 	#[allow(deprecated)]
 	let res = runtime_api.function_signature_changed_before_version_2(&block_id).unwrap();
 	assert_eq!(&res, &[1, 2]);
 }
+
+#[test]
+#[should_panic]
+fn calling_with_both_strategy_and_fail_on_wasm_should_panic() {
+	let client = test_client::new_with_execution_strategy(ExecutionStrategy::Both);
+	let runtime_api = client.runtime_api();
+	let block_id = BlockId::Number(client.info().unwrap().chain.best_number);
+	assert_eq!(runtime_api.fail_on_wasm(&block_id).unwrap(), 1);
+}
+
+#[test]
+fn calling_with_both_strategy_and_fail_on_native_should_work() {
+	let client = test_client::new_with_execution_strategy(ExecutionStrategy::Both);
+	let runtime_api = client.runtime_api();
+	let block_id = BlockId::Number(client.info().unwrap().chain.best_number);
+	assert_eq!(runtime_api.fail_on_native(&block_id).unwrap(), 1);
```

Comment:
Shouldn't this fail as well? As both are executed and return a different result?

Reply:
I think that in this case [this handler](https://github.com/paritytech/substrate/blob/9151615579cf9f92e6dfd1497f69d3a2281b1fef/core/state-machine/src/lib.rs#L264:L270) is triggered and then the wasm result is returned.